src/back/spv/writer.rs

```diff
@@ -450,147 +838,92 @@ impl Writer {
 
         match constant.inner {
             crate::ConstantInner::Sint(val) => {
-                let type_id = self.get_type_id(arena, constant.ty);
-
-                let mut instruction = Instruction::new(Op::Constant);
-                instruction.set_type(type_id);
-                instruction.set_result(id);
-
                 let ty = &ir_module.types[constant.ty];
-                match ty.inner {
+                let values = match ty.inner {
                     crate::TypeInner::Scalar { kind: _, width } => match width {
-                        32 => {
-                            instruction.add_operand(val as u32);
-                        }
+                        32 => vec![val as u32],
                         64 => {
                             let (low, high) = ((val >> 32) as u32, val as u32);
-                            instruction.add_operand(low);
-                            instruction.add_operand(high);
+                            vec![low, high]
                         }
                         _ => unreachable!(),
                     },
                     _ => unreachable!(),
-                }
+                };
 
+                let type_id = self.get_type_id(arena, constant.ty);
+                let instruction = self.instruction_constant(type_id, id, values);
                 (instruction, id)
             }
             crate::ConstantInner::Uint(val) => {
-                let type_id = self.get_type_id(arena, constant.ty);
-
-                let mut instruction = Instruction::new(Op::Constant);
-                instruction.set_type(type_id);
-                instruction.set_result(id);
-
                 let ty = &ir_module.types[constant.ty];
-                match ty.inner {
+                let values = match ty.inner {
                     crate::TypeInner::Scalar { kind: _, width } => match width {
-                        32 => {
-                            instruction.add_operand(val as u32);
-                        }
+                        32 => vec![val as u32],
```

Comment:
nit: in cases like this you can avoid heap allocation by just returning slices, i.e. `&[val as u32][..]`